File: python/P032_LongestValidParentheses/LongestValidParentheses.py
# -*- coding: utf-8 -*-
class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        stack = [-1]
        ret = 0
        for i in range(len(s)):
            if s[i] == ')' and len(stack) > 1 and s[stack[-1]] == '(':
                stack.pop()
                ret = max(ret, i-stack[-1])
            else:
                stack.append(i)
        return ret

File: java/src/P032/LongestValidParentheses/LongestValidParentheses.java
package P032.LongestValidParentheses;

import java.util.ArrayDeque;
import java.util.Deque;

public class LongestValidParentheses {
    public int longestValidParentheses(String s) {
        Deque<Integer> stack = new ArrayDeque<>();
        int ret = 0;
        stack.push(-1);
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == ')' && stack.size() > 1 && s.charAt(stack.peek()) == '(') {
                stack.pop();
                ret = Math.max(ret, i - stack.peek());
            } else {
                stack.push(i);
            }
        }
        return ret;
    }
}
